libs/mrwebrtc/src/interop/remote_audio_track_interop.cpp

```diff
@@ -1,72 +1,123 @@
+#include "utils.h"
+
+using namespace Microsoft::MixedReality::WebRTC;
+
+MRS_API void MRS_CALL
+mrsRemoteAudioTrackSetUserData(mrsRemoteAudioTrackHandle handle,
+                               void* user_data) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(handle)) {
+    track->SetUserData(user_data);
+  }
+}
+
+MRS_API void* MRS_CALL
+mrsRemoteAudioTrackGetUserData(mrsRemoteAudioTrackHandle handle) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(handle)) {
+    return track->GetUserData();
+  }
+  return nullptr;
+}
+
+void MRS_CALL
+mrsRemoteAudioTrackRegisterFrameCallback(mrsRemoteAudioTrackHandle trackHandle,
+                                         mrsAudioFrameCallback callback,
+                                         void* user_data) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(trackHandle)) {
+    track->SetCallback(AudioFrameReadyCallback{callback, user_data});
+  }
+}
+
+mrsResult MRS_CALL
+mrsRemoteAudioTrackSetEnabled(mrsRemoteAudioTrackHandle track_handle,
+                              mrsBool enabled) noexcept {
+  auto track = static_cast<RemoteAudioTrack*>(track_handle);
+  if (!track) {
+    return Result::kInvalidParameter;
+  }
+  track->SetEnabled(enabled != mrsBool::kFalse);
+  return Result::kSuccess;
+}
+
+mrsBool MRS_CALL
+mrsRemoteAudioTrackIsEnabled(mrsRemoteAudioTrackHandle track_handle) noexcept {
+  auto track = static_cast<RemoteAudioTrack*>(track_handle);
+  if (!track) {
+    return mrsBool::kFalse;
+  }
+  return (track->IsEnabled() ? mrsBool::kTrue : mrsBool::kFalse);
+}
+
+void MRS_CALL
+mrsRemoteAudioTrackOutputToDevice(mrsRemoteAudioTrackHandle track_handle,
+    bool output) noexcept {
+  if (auto track = static_cast<RemoteAudioTrack*>(track_handle)) {
+      track->OutputToDevice(output);
+  }
+}
+
+mrsBool MRS_CALL mrsRemoteAudioTrackIsOutputToDevice(
+    mrsRemoteAudioTrackHandle track_handle) noexcept {
+  if (auto track = static_cast<const RemoteAudioTrack*>(track_handle)) {
+    return track->IsOutputToDevice() ? mrsBool::kTrue : mrsBool::kFalse;
+  }
+  return mrsBool::kFalse;
+}
+
+mrsResult MRS_CALL
+mrsRemoteAudioTrackCreateReadBuffer(mrsRemoteAudioTrackHandle track_handle,
+                              mrsAudioTrackReadBufferHandle* audioBufferOut) {
+  if (!audioBufferOut) {
+      return Result::kInvalidParameter;
+  }
+
+  *audioBufferOut = nullptr;
+  if (auto track = static_cast<RemoteAudioTrack*>(track_handle)) {
+    auto buffer = track->CreateReadBuffer();
+    *audioBufferOut = buffer.release();
+    return Result::kSuccess;
+  }
+  return Result::kInvalidNativeHandle;
+}
+
+mrsResult MRS_CALL
+mrsAudioTrackReadBufferRead(mrsAudioTrackReadBufferHandle buffer,
+                            int sample_rate,
+                            int num_channels,
+                            mrsAudioTrackReadBufferPadBehavior pad_behavior,
+                            float* data_out,
+                            int* data_len_in_out,
```

Comment:
This is out only in the header. Even though longer, perhaps separate in & out size params are more idiomatic?

Reply:
We use in+out parameters in a few places already, so it's more consistent like this.

Reply:
The header and docstring is inconsistent.

```
mrsAudioTrackReadBufferRead(mrsAudioTrackReadBufferHandle buffer,
                            ...
                            float* data_out,
                            int* num_read_samples_out,
```

Reply:
BTW I agree inout combining is OK here since the API is internal and the param list is long.

Reply:
Yes it was meant to be `num_read_samples_in_out` and the docs are incomplete - I copied the C# comments where the buffer size is passed in the array, and forgot to update 😄 

This is public API technically, and I agree separate in and out make sense (they mean different things) so will separate.

Reply:
Fixed